diff --git a/dist/index.cjs b/dist/index.cjs
index 9c4a059012663dff884e5d26876c692e072b36f3..de3499170b8beb0c25419146fefca0b6c5b8aa4d 100644
--- a/dist/index.cjs
+++ b/dist/index.cjs
@@ -64,7 +64,9 @@ var EventManager = class {
   }
   off(event, listener) {
     const listeners = this._eventListeners.get(event) ?? [];
-    const index = listeners.indexOf(listener);
+    const index = listeners.findIndex(
+      (v) => v === listener || (v._originalListener === listener && v._originalListener !== undefined)
+    );
     if (index !== -1) {
       listeners.splice(index, 1);
     }
@@ -77,6 +79,7 @@ var EventManager = class {
       listener(...arguments_);
       this.off(event, onceListener);
     };
+    onceListener._originalListener = listener;
     this.on(event, onceListener);
   }
   // Emit an event
diff --git a/dist/index.d.cts b/dist/index.d.cts
index 8d20f2cc2a6b2c431cc491e3a754deeb5a6fae72..eb151fb8bb798c39551aae9d8cfc0738dae87d74 100644
--- a/dist/index.d.cts
+++ b/dist/index.d.cts
@@ -1,4 +1,6 @@
-type EventListener = (...arguments_: any[]) => void;
+type EventListener = ((...arguments_: any[]) => void) & {
+  _originalListener?: EventListener
+};
 declare class EventManager {
     _eventListeners: Map<string, EventListener[]>;
     _maxListeners: number;
diff --git a/dist/index.d.ts b/dist/index.d.ts
index 8d20f2cc2a6b2c431cc491e3a754deeb5a6fae72..eb151fb8bb798c39551aae9d8cfc0738dae87d74 100644
--- a/dist/index.d.ts
+++ b/dist/index.d.ts
@@ -1,4 +1,6 @@
-type EventListener = (...arguments_: any[]) => void;
+type EventListener = ((...arguments_: any[]) => void) & {
+  _originalListener?: EventListener
+};
 declare class EventManager {
     _eventListeners: Map<string, EventListener[]>;
     _maxListeners: number;
diff --git a/dist/index.js b/dist/index.js
index 0b8687711548a42cdc3c0fcb11ca2696eb4e1aed..8ded3010676444127f748f51a0c6ba4154c4ed37 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -34,7 +34,9 @@ var EventManager = class {
   }
   off(event, listener) {
     const listeners = this._eventListeners.get(event) ?? [];
-    const index = listeners.indexOf(listener);
+    const index = listeners.findIndex(
+      (v) => v === listener || (v._originalListener === listener && v._originalListener !== undefined)
+    );
     if (index !== -1) {
       listeners.splice(index, 1);
     }
@@ -47,6 +49,7 @@ var EventManager = class {
       listener(...arguments_);
       this.off(event, onceListener);
     };
+    onceListener._originalListener = listener;
     this.on(event, onceListener);
   }
   // Emit an event
